README.md
# Game of Life
A Javascript implementation of [Conway's Game of Life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life).
>v1.1.0

* Implemented with ES6 Javascript and HTML5 <canvas> element.
* No dependencies, except for [canvas support](http://caniuse.com/#feat=canvas).
* It's possible to run multiple instances on the same page.
* Set an arbitrary state using the setState() method.

## Installation
1. `bower install --save-dev gameoflife`, `npm install --save-dev gameoflife-es6` or download `/dist/gameoflife.min.js`.
2. Add a canvas-element to your html with an ID of your choice.  
3. Include the following Javascript in your html file, just before closing the body tag:
 ```
    <script src="src/gameoflife.js"></script>
    <script type="text/javascript">
        var myGameOfLife = new GameOfLife();
    </script>
``` 
4. Add some buttons to control the game. See our [example page](http://htmlpreview.github.io/?https://github.com/barryvanveen/gameoflife/blob/master/example.html) and check out the Usage section below.

>If you are using ES6 you can also import the GameOfLife class from /src/gameoflife.js

## Usage
Initialise a new instance using `var myGame = new GameOfLife();`. Override default parameters using `var myCustomGame
 = new GameOfLife({canvas_id: "myCustomID", num_cols: 123})`.
 
Click on cells in the grid to change their state. You can also set the state with the setState() method. Once your 
finished providing the initial state, call the start() method. 
 
###setState(cells)

Provide an array of cells. Each cell is an object consisting of the column and row, eg. `{col: 1, row: 2}`. The state
 of each cell is changed but no further checks are made. 
 
Example of adding a [Blinker](http://www.conwaylife.com/wiki/Blinker) pattern to an empty state:
```
MyGame.setState([
    {col: 1, row: 0},
    {col: 1, row: 1},
    {col: 1, row: 2}
]);
```

### start()
Starts iterating of states of the game. The game stops automatically if the game ends up in dead state, ie. no 
changes are made by further iterations. 

### step()
Compute the next state of the population. If the game is currently running it is stopped after this iteration. 

### stop()
Stop the game from iterating any further.

### reset()
Reset to an empty state. If the game is currently running it is stopped.

## Configuration
Below is a list of options with the default values.

* `canvas_id` (String, "gameoflife_canvas") ID that identifies the canvas element. 
* `num_cols` (Int, 80) Number of columns in the grid.  
* `num_rows` (Int, 40) Number of rows in the grid.
* `cell_size` (Int, 10) Size of each cell, in pixels.
* `color_lines` (String, "#cccccc") Color of the grid lines.
* `color_cell_dead` (String, "#ffffff") Color of dead cells.
* `color_cell_alive` (String, "#57A0DB") Color of live cells.
* `update_interval` (Int, 50) Update interval in milliseconds.

## Examples
Take a look at the [example page](http://htmlpreview.github.io/?https://github.com/barryvanveen/gameoflife/blob/master/example.html).

## Licence
This software is freely distributable under the terms of the [MIT license](LICENCE).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
### Added
### Changed
### Deprecated
### Removed
### Fixed

## [1.1.0] - 2016-06-25
### Changed
- Changed gulp workflow for rollup integration

## [1.0.3] - 2016-06-06
### Fixed
- The _buildConfig method now always returns a config object 

## [1.0.2] - 2016-06-06
### Added
- Added npm installation command to readme  
### Changed
- Changed npm package name to "gameoflife-es6"

## [1.0.1] - 2016-06-04
### Fixed
- Fixed bower.json so that this project is not marked as private

## [1.0.0] - 2016-06-04
### Added
- First release


gulpfile.js
var babel = require("gulp-babel");
var gulp = require('gulp');
var rename = require("gulp-rename");
var rollup = require('rollup-stream');
var sourcemaps = require('gulp-sourcemaps');
var source = require('vinyl-source-stream');
var buffer = require('vinyl-buffer');

/**
 * concatenate and minify js files and write a sourcemap
 */
gulp.task('build-js', function () {
    rollup({
            entry: './src/gameoflife.js',
            sourceMap: true,
            format: 'iife',
            moduleName: 'GameOfLife',
            banner: `//  GameOfLife JavaScript Plugin v1.1.0
//  https://github.com/barryvanveen/gameoflife
//
//  Released under the MIT license
//  http://choosealicense.com/licenses/mit/`
        })

        // print errors to console, this makes sure Gulp can keep watching and running
        .on('error', e => {
            console.error(`${e.stack}`);
        })

        // point to the entry file
        .pipe(source('gameoflife.js', './src'))

        // buffer the output, needed for gulp-sourcemaps
        .pipe(buffer())

        // tell gulp-sourcemaps to load the inline sourcemap produced by rollup-stream
        .pipe(sourcemaps.init({loadMaps: true}))

        // compile es2015 into plain javascript
        .pipe(babel({presets: ['es2015']}))

        // rename output file
        .pipe(rename('gameoflife.min.js'))

        // write source map
        .pipe(sourcemaps.write('.'))

        // output to /dist
        .pipe(gulp.dest('dist'));
});

/**
 * watch for changes in js-files, then build-js
 */
gulp.task('watch-js', function(){
    gulp.watch('src/**/*.js', ['build-js']);
});

/**
 * perform these tasks when running just 'gulp'
 */
gulp.task('default', ['build-js', 'watch-js']);

src/state.js
import Helpers from './helpers';

export default class State {

    constructor(config) {

        this.config = config;

        var col = this.config.num_cols,
            row;

        this.cells = [];
        this.changes = [];

        while (col--) {
            this.cells[col] = [];

            row = this.config.num_rows;
            while (row--) {
                this.cells[col][row] = 0;
            }
        }

    };

    changeCell(col, row) {
        this.cells[col][row] = !this.cells[col][row];

        return this.cells[col][row];
    };

    computeNextState() {

        var count = 0,
            col, row, rowOffset, colOffset, neighborCol, neighborRow;

        this.changes = [];

        // iterate over all cells
        col = this.config.num_cols;
        while (col--) {

            row = this.config.num_rows;
            while (row--) {

                count = 0;

                // iterate over all neighbors in Moore neighborhood with radius=1
                for (colOffset = -1; colOffset <= 1; ++colOffset) {
                    for (rowOffset = -1; rowOffset <= 1; ++rowOffset) {
                        if (colOffset == 0 && rowOffset == 0) {
                            continue;
                        }

                        neighborCol = col + colOffset;
                        if (neighborCol < 0 || neighborCol >= this.config.num_cols) {
                            neighborCol = Helpers.mod(this.config.num_cols, neighborCol);
                        }

                        neighborRow = row + rowOffset;
                        if (neighborRow < 0 || neighborRow >= this.config.num_rows) {
                            neighborRow = Helpers.mod(this.config.num_rows, neighborRow);
                        }

                        // count neighbors that are "on" or "alive"
                        if (this.cells[neighborCol][neighborRow]) {
                            count++;
                        }
                    }
                }

                // determine state of new cells
                // die if under- or overpopulated
                if ((count < 2 || count > 3) && this.cells[col][row]) {
                    this.changes.push({col, row});
                // come to life if exactly 3 neighbors
                } else if (count == 3 && !this.cells[col][row]) {
                    this.changes.push({col, row});
                }

            }
        }

        return this.changes;

    };

}


src/canvas.js
import Helpers from './helpers';

export default class Canvas {

    constructor(config, state) {

        this.config = config;
        this.state = state;

        this.canvas = document.getElementById(this.config.canvas_id);

        if (this.canvas == null) {
            throw new Error("Canvas element could not be found.");
        }

        this.context = this.canvas.getContext("2d");

        if (!this.context) {
            throw new Error("Canvas context could not be retrieved.");
        }

        this.canvas.width = (this.config.num_cols * this.config.cell_size) + 1;
        this.canvas.height = (this.config.num_rows * this.config.cell_size) + 1;

        // clear the canvas
        this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);

        // fill the background
        this.context.fillStyle = this.config.color_cell_dead;
        this.context.strokeStyle = this.config.color_lines;
        this.context.fillRect(1, 1, this.canvas.width-1,  this.canvas.height-1);

        // vertical lines
        for (var x = 0; x <= this.canvas.width; x += this.config.cell_size) {
            this.context.moveTo(0.5 + x, 0);
            this.context.lineTo(0.5 + x, this.canvas.width);
        }

        // horizontal lines
        for (var y = 0; y <= this.canvas.width; y += this.config.cell_size) {
            this.context.moveTo(0, 0.5 + y);
            this.context.lineTo(this.canvas.width, 0.5 + y);
        }

        // draw it
        this.context.stroke();

    };

    _handleClick(e) {

        var cell = this._getCellFromCursorPosition(e);

        if (cell == false) {
            return;
        }

        var cellState = this.state.changeCell(cell[0], cell[1]);

        this._drawCell(cell[0], cell[1], cellState);

    };

    _getCellFromCursorPosition(e) {

        var left, top;

        // get coordinates of click on page
        if (typeof(e.pageX) != "undefined" && typeof(e.pageY) != "undefined") {
            left = e.pageX;
            top = e.pageY;
        } else {
            left = e.clientX + document.body.scrollLeft + document.documentElement.scrollLeft;
            top = e.clientY + document.body.scrollTop + document.documentElement.scrollTop;
        }

        // get coordinates relative to canvas
        var canvas_offset = Helpers.getPosition(this.canvas);
        left -= canvas_offset[0];
        top -= canvas_offset[1];

        if (left > (this.config.num_cols * this.config.cell_size) || top > (this.config.num_rows * this.config.cell_size)) {
            return false;
        }

        // now calculate in which cell this click falls
        return [Math.floor(left / this.config.cell_size), Math.floor(top / this.config.cell_size)];

    };

    _drawCell(col, row, cellState) {

        if (cellState) {
            this.context.fillStyle = this.config.color_cell_alive;
        } else {
            this.context.fillStyle = this.config.color_cell_dead;
        }

        // fill rectangle from (col-1,row-1) with width and height of cellSize-1
        this.context.fillRect(1 + (col*this.config.cell_size), 1 + (row*this.config.cell_size), this.config.cell_size-1, this.config.cell_size-1);

    };

}


src/example.js
import GameOfLife from './gameoflife';

var BigGame = new GameOfLife({
    canvas_id:              "gameoflife_canvas",
    num_cols:               80,
    num_rows:               40,
    cell_size:              10,
    color_lines:            "#cccccc",
    color_cell_empty:       "#ffffff",
    color_cell_selected:    "#57A0DB",
    update_interval:        50
});

$('.game1_start').click(function() {BigGame.start()});
$('.game1_stop').click(function() {BigGame.stop()});
$('.game1_step').click(function() {BigGame.step()});
$('.game1_reset').click(function() {BigGame.reset()});

var SmallGame = new GameOfLife({
    canvas_id:              "gameoflife_canvas_small",
    num_cols:               80,
    num_rows:               40,
    cell_size:              5,
    color_lines:            "#cccccc",
    color_cell_empty:       "#ffffff",
    color_cell_selected:    "#57A0DB",
    update_interval:        50
});

$('.game2_start').click(function() {SmallGame.start()});
$('.game2_stop').click(function() {SmallGame.stop()});
$('.game2_step').click(function() {SmallGame.step()});
$('.game2_reset').click(function() {SmallGame.reset()});

src/gameoflife.js
import State from './state';
import Canvas from './canvas';

export default class GameOfLife {

    constructor(customConfig) {
        this._defaults = {
            canvas_id:              "gameoflife_canvas",
            num_cols:               80,
            num_rows:               40,
            cell_size:              10,
            color_lines:            "#cccccc",
            color_cell_dead:        "#ffffff",
            color_cell_alive:       "#57A0DB",
            update_interval:        50
        };
        this._interval = null;

        this.config = this._buildConfig(customConfig);
        this.state = new State(this.config);
        this.canvas = new Canvas(this.config, this.state);
        this._setEventListeners();
    };

    _buildConfig(customConfig) {

        var i;

        var config = this._defaults;

        if (typeof(customConfig) != "object") {
            return config;
        }

        for (i in customConfig) {
            if (typeof(config[i]) == "undefined" || typeof(customConfig[i]) == "object") {
                continue;
            }
            config[i] = customConfig[i];
        }

        return config;

    };

    _setEventListeners() {

        var self = this;

        this.canvas.canvas.addEventListener('click', function(e) {
            self.canvas._handleClick(e);
        }, false);

    };

    setState(cells) {

        var i, cellState;

        this.stop();

        if (cells.length == 0) {
            return;
        }

        for (i=0; i<cells.length; i++) {
            cellState = this.state.changeCell(cells[i].col, cells[i].row);
            this.canvas._drawCell(cells[i].col, cells[i].row, cellState);
        }

    }

    reset() {

        this.stop();

        this.state = new State(this.config);
        this.canvas = new Canvas(this.config, this.state);

    };

    start() {

        if (this._interval != null) {
            return;
        }

        var self = this,
            i, cellState;

        this._interval = setInterval(function() {
            var changes = self.state.computeNextState();

            if (changes.length == 0) {
                self.stop();
                return;
            }

            for (i=0; i<changes.length; i++) {
                cellState = self.state.changeCell(changes[i].col, changes[i].row);
                self.canvas._drawCell(changes[i].col, changes[i].row, cellState);
            }
        }, this.config.update_interval);

    };

    step() {

        var changes, cellState, i;

        this.stop();

        changes = this.state.computeNextState();

        if (changes.length == 0) {
            return;
        }

        for (i=0; i<changes.length; i++) {
            cellState = this.state.changeCell(changes[i].col, changes[i].row);
            this.canvas._drawCell(changes[i].col, changes[i].row, cellState);
        }

    };

    stop() {

        clearInterval(this._interval);
        this._interval = null;

    };

}
